src/Microsoft.AspNetCore.SystemWebAdapters.Authentication/RedirectUrlProcessor.cs

```diff
@@ -0,0 +1,68 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.WebUtilities;
+using Microsoft.Extensions.Primitives;
+
+namespace Microsoft.AspNetCore.SystemWebAdapters.Authentication;
+
+/// <summary>
+/// Processes remote authentication results to fix-up 'ReturnUrl' parameters
+/// given in redirect responses so that they will return the user to the correct
+/// original URL instead of to the remote authentication endpoint.
+/// </summary>
+internal class RedirectUrlProcessor : IRemoteAuthenticationResultProcessor
+{
+    private const string LocationHeaderName = "Location";
+    private const string ReturnUrlQueryStringName = "ReturnUrl";
+
+    /// <summary>
+    /// Updates response headers in a remote authentication result so that any
+    /// 'Location' headers will include a ReturnUrl query string corresponding to the
+    /// URL of the original request rather than the URL of the remote authentication endpoint.
+    /// </summary>
+    /// <param name="result"></param>
+    /// <param name="context"></param>
+    /// <returns></returns>
+    public Task ProcessAsync(RemoteAuthenticationResult result, HttpContext context)
+    {
+        if (result.ResponseHeaders.TryGetValue(LocationHeaderName, out var locationHeaders))
```

Comment:
Set-Cookie headers can have a similar problem with their path and domain fields.